### ISSUE TITLE
How to design a serial port addon?

### ISSUE THREAD

**Issue by Risca** [Zenterio/opensourcelib#6]
I've been planning on creating a k2-serial addon, but I haven't come up with how to design it. My thoughts so far:

* The extension would use the message bus to send lines as they appear on the serial port
* The log lines will also be saved to a log file using the logger framework
* The serial port has to be connected to a SUT, so that the same test can be run on multiple SUTs, using unique serial ports
* It should be possible to do `zk2 serial --port /dev/ttyUSB0 send "hello world!"`

So far, so good. Nothing tricky about it. I guess some helper components that wrap some of the details would be needed to make the test cases look nice and easy.

Now comes the more tricky part:

* I want to be able to test the interaction between 2 (or more) SUTs in a single testcase
* Each SUT can have multiple serial ports to control it. E.g.
  1. debug log and terminal
  1. AT interface
 
How would that kind of configuration be designed?
Like this?
```
zk2 --suts-ids one --suts-one@serial-ids debug --suts-one@serial-ids at --suts-one@serial-debug-port /dev/ttyUSB0 --suts-one@serial-at-port /dev/ttyUSB1 <repeat for SUT two>
```

Maybe create multiple extensions called `DualSerial`, `TripleSerial`, etc.?

Bonus points:
* the serial port might be located on another device and a TCP socket is used to access it. I.e.:
    ```
    k2 host <-tcp-> ser2net <-rs232-> SUT
    ```

_For an initial extension, I think it's fine to only support 1 (physical) serial port connected to a SUT._

I would appreciate some feedback on how to do the code separation here. Where do I place my call to `serial.Serial()`? The serial port is an exclusive resource so I have to avoid making multiple calls to `Serial()`, but at the same time, multiple components might need access to the port. I guess the message bus would solve that problem at least one way (reading lines from the port).
Should I design some kind of abstract interface, similar to the powermeter and powerswitch extensions?

**Comment by per-bohlin** [comment]
I've added two addons from upstream code base, one telnet and one serial (#8). Hopefully they can be of use.
They only support single configuration and assumes direct communication between K2 host and SUT.
Feel free to expand them to support multiple configurations.

The case of k2 talking serial with SUT across TCP adapter will require a new addon. The general structure of the serial addon can serve as an example of how to reserve the device for a given execution context - to avoid cross talk between different components  trying to access it at the same time. So I would recommend to build  a similar structure for the TCP-adapter based variant - assuming that the "ser2net" adapter doesn't handle it for you. If it does, then perhaps the multi-session telnet might even work out of the box - assuming the ser2net talks telnet. At least the new plugin can be modeled similar to the telnet plugin in that case.

**Comment by Risca** [comment]
`ser2net` doesn't talk telnet per se, it depends if the actual serial port on the other end is connected to a terminal that talks telnet, but the telnet component should probably work in such cases.

Thanks a lot for these latest updates! :smiley_cat:

**Comment by per-bohlin** [comment]
Do you want further feedback on this topic or can we close the issue?

**Comment by Risca** [comment]
I'm still not sure how to structure an addon that allow multiple serial ports for the same SUT. Maybe a hard coded extension for dual serial ports, and another for three serial ports, etc. (as need arises). Example config would then be:

```yaml
suts.ids: chip
suts.chip.dualserial.enabled: True
suts.chip.dualserial.first.port: /dev/ttyUSB0
suts.chip.dualserial.second.port: /dev/ttyUSB0
```

I would be nice to have a generic extension, but maybe it's not worth the effort/complexity?

Maybe a bit off topic for this issue, but how do I write a test case that uses multiple suts? Is that what the `uses` argument is for?

I think those are the two remaining questions that I have before we can close this issue. I believe I can solve the "tcp socket serial port" issue myself =)

**Comment by per-bohlin** [comment]
I will look into how to best do a multi-port serial addon configuration and get back to you. I'm quite sure it can be done. I just don't remember exactly how and need to dig into it a bit.

**Comment by per-bohlin** [comment]
See k2-infra/virt/docker.py, it uses the techniques needed  with DOCKER_IMAGES_IDS and DOCKER_IMAGES_HOSTNAME for multi-connection per SUT.
I will followup with some conceptual code example to illustrate the point further.

**Comment by per-bohlin** [comment]
Example code (might not work, have not tested it):
```python
from zaf.config.options import ConfigOptionId

from k2.sut import SUT

SERIAL_CONNECTION_IDS = ConfigOptionId(
    'serial.ids',
    'Identifies a serial connection',
    namespace='serial.connections',
    entity=True,
    multiple=True,
)

SERIAL_CONNECTION_DEVICE = ConfigOptionId(
    'device',
    'The serial device. Multiple formats supported e.g [ttyXXX, /dev/ttyXXX, 3-10.5.1, guess]',
    at=SERIAL_CONNECTION_IDS
)

SERIAL_CONNECTION_BAUDRATE = ConfigOptionId(
    'baudrate',
    'The baudrate of the serial connection',
    at=SERIAL_CONNECTION_IDS,
    option_type=int,
    default=115200
)

SUT_SERIAL_CONNECTIONS = ConfigOptionId(
    'serial.connections',
    'Serial connections available for this sut',
    at=SUT,
    multiple=True,
)
```
Associated example config:
```yaml

sut.ids:
    - mysut

serial.connections.ids:
    - con_a
    - con_b

serial.connections.con_a.device: /dev/ttyXXA
serial.connections.con_b.device: /dev/ttyXXB

suts.mysut.serial.connections:
    - con_a
    - con_b
```

The example only covers two properties, device and baudrate. Each property for the connection should then follow the same pattern.

To make it backward compatible, the existing configuration could apply to a "default" connection that is used unless specified.
The serial connection component could then take an optional argument for which connection to use, which would default to "default".

**Comment by Risca** [comment]
That sounds like a nice approach. I think I have enough info to get started now :smiley:
